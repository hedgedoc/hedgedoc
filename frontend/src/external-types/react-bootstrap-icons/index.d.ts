/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

declare module 'react-bootstrap-icons/dist/icons/*' {
  import type { Icon } from 'react-bootstrap-icons'
  const ReactComponent: Icon
  export default ReactComponent
}
