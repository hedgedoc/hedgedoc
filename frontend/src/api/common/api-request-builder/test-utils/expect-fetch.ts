/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { defaultConfig } from '../../default-config'
import { Mock } from 'ts-mockery'

/**
 * Mock fetch api for tests.
 * Check that the given url and options are present in the request and return the given status code.
 *
 * @param expectedUrl the url that should be requested
 * @param requestStatusCode the status code the mocked request should return
 * @param expectedOptions additional options
 */
export const expectFetch = (
  expectedUrl: string,
  requestStatusCode: number,
  expectedOptions: RequestInit,
  responseBody?: unknown
): void => {
  global.fetch = jest.fn((fetchUrl: RequestInfo | URL, fetchOptions?: RequestInit): Promise<Response> => {
    expect(fetchUrl).toEqual(expectedUrl)
    expect(fetchOptions).toStrictEqual({
      ...defaultConfig,
      body: undefined,
      headers: new Headers(),
      ...expectedOptions
    })
    return Promise.resolve(
      Mock.of<Response>({
        status: requestStatusCode,
        statusText: mapCodeToText(requestStatusCode),
        json: jest.fn(() => (responseBody ? Promise.resolve(responseBody) : Promise.reject()))
      })
    )
  }) as typeof global.fetch
}
const mapCodeToText = (code: number): string => {
  switch (code) {
    case 400:
      return 'bad_request'
    case 401:
      return 'forbidden'
    default:
      return 'unknown_code'
  }
}
