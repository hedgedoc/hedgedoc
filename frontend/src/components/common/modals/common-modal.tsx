/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { concatCssClasses } from '../../../utils/concat-css-classes'
import type { PropsWithDataCypressId } from '../../../utils/cypress-attribute'
import { cypressId } from '../../../utils/cypress-attribute'
import { testId } from '../../../utils/test-id'
import { UiIcon } from '../icons/ui-icon'
import { ShowIf } from '../show-if/show-if'
import type { PropsWithChildren, ReactElement } from 'react'
import React, { Fragment, useMemo } from 'react'
import { Modal } from 'react-bootstrap'
import type { Icon } from 'react-bootstrap-icons'
import { Trans, useTranslation } from 'react-i18next'

export interface ModalVisibilityProps {
  show: boolean
  onHide?: () => void
}

export interface ModalContentProps {
  titleI18nKey?: string
  title?: string
  showCloseButton?: boolean
  titleIcon?: Icon
  modalSize?: 'lg' | 'sm' | 'xl'
  additionalClasses?: string
  additionalTitleElement?: ReactElement
}

export type CommonModalProps = PropsWithDataCypressId & ModalVisibilityProps & ModalContentProps

/**
 * Renders a generic modal.
 *
 * @param show If the modal should be shown or not.
 * @param onHide The callback to hide the modal again
 * @param title The title in the header of the modal
 * @param showCloseButton If a close button should be shown
 * @param titleIcon An optional title icon
 * @param additionalClasses Additional class names for the modal
 * @param modalSize The modal size
 * @param children The children to render into the modal.
 * @param titleIsI18nKey If the title is a i18n key and should be used as such
 * @param additionalTitleElement additional optional element that should be shown in the header
 * @param props Additional props directly given to the modal
 */
export const CommonModal: React.FC<PropsWithChildren<CommonModalProps>> = ({
  show,
  onHide,
  titleI18nKey,
  title,
  showCloseButton,
  titleIcon,
  additionalClasses,
  modalSize,
  children,
  additionalTitleElement,
  ...props
}) => {
  useTranslation()

  const titleElement = useMemo(() => {
    return titleI18nKey !== undefined ? <Trans i18nKey={titleI18nKey} /> : <span>{title}</span>
  }, [titleI18nKey, title])

  return (
    <ShowIf condition={show}>
      <Modal
        {...cypressId(props)}
        show={show}
        onHide={onHide}
        animation={true}
        {...testId('commonModal')}
        dialogClassName={concatCssClasses(additionalClasses)}
        size={modalSize}>
        <Modal.Header closeButton={!!showCloseButton}>
          <Modal.Title>
            <UiIcon icon={titleIcon} nbsp={true} />
            {titleElement}
          </Modal.Title>
          {additionalTitleElement ?? <Fragment />}
        </Modal.Header>
        {children}
      </Modal>
    </ShowIf>
  )
}
