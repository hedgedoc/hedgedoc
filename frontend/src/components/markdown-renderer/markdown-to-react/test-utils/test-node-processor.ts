/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { NodeProcessor } from '../../node-preprocessors/node-processor'
import type { Document } from 'domhandler'
import { Element, Text } from 'domhandler'

export class TestNodeProcessor extends NodeProcessor {
  process(document: Document): Document {
    document.childNodes.push(new Element('node-processor', {}, [new Text('node processor children')]))
    return document
  }
}
