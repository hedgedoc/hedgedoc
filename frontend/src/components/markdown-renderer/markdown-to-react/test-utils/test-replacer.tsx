/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { NativeRenderer, NodeReplacement, SubNodeTransform } from '../../replace-components/component-replacer'
import { ComponentReplacer, DO_NOT_REPLACE } from '../../replace-components/component-replacer'
import type { Element } from 'domhandler'
import React, { Fragment } from 'react'

export class TestReplacer extends ComponentReplacer {
  replace(node: Element, subNodeTransform: SubNodeTransform, nativeRenderer: NativeRenderer): NodeReplacement {
    return node.tagName === 'node-processor' ? (
      <Fragment>
        <span>NodeProcessor! </span>
        <span data-children={true}> {node.childNodes.map(subNodeTransform)} </span>
        <span data-native={true}> {nativeRenderer()} </span>
      </Fragment>
    ) : (
      DO_NOT_REPLACE
    )
  }
}
