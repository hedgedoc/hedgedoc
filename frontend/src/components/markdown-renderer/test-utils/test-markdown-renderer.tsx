/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { StoreProvider } from '../../../redux/store-provider'
import type { MarkdownRendererExtension } from '../extensions/_base-classes/markdown-renderer-extension'
import { MarkdownToReact } from '../markdown-to-react/markdown-to-react'
import React, { useMemo } from 'react'

export interface SimpleMarkdownRendererProps {
  content: string
  extensions: MarkdownRendererExtension[]
}

/**
 * A markdown renderer for tests.
 *
 * @param content The content to be rendered.
 * @param extensions The {@link MarkdownRendererExtension MarkdownExtensions} to use for rendering.
 */
export const TestMarkdownRenderer: React.FC<SimpleMarkdownRendererProps> = ({ content, extensions }) => {
  const lines = useMemo(() => content.split('\n'), [content])
  return (
    <StoreProvider>
      <MarkdownToReact
        markdownContentLines={lines}
        markdownRenderExtensions={extensions}
        newlinesAreBreaks={true}
        allowHtml={false}
      />
    </StoreProvider>
  )
}
