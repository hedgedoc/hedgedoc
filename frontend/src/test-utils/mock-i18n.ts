/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { TFunction } from 'i18next'
import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'

/**
 * Initializes i18n with minimal settings and without any data, so it just returns the used key as translation.
 *
 * @return A promise that resolves if i18n has been initialized
 */
export const mockI18n = (): Promise<TFunction> => {
  return i18n.use(initReactI18next).init({
    lng: 'en',
    fallbackLng: 'en',
    ns: ['translationsNS'],
    defaultNS: 'translationsNS',
    interpolation: {
      escapeValue: false
    },
    resources: { en: { translationsNS: {} } }
  })
}
