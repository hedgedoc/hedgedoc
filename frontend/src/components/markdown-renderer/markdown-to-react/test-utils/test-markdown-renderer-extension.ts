/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { MarkdownRendererExtension } from '../../extensions/_base-classes/markdown-renderer-extension'
import type { NodeProcessor } from '../../node-preprocessors/node-processor'
import type { ComponentReplacer } from '../../replace-components/component-replacer'
import { TestNodeProcessor } from './test-node-processor'
import { TestReplacer } from './test-replacer'
import type MarkdownIt from 'markdown-it'
import Token from 'markdown-it/lib/token'

export class TestMarkdownRendererExtension extends MarkdownRendererExtension {
  buildNodeProcessors(): NodeProcessor[] {
    return [new TestNodeProcessor()]
  }

  configureMarkdownIt(markdownIt: MarkdownIt) {
    markdownIt.use(() => {
      markdownIt.core.ruler.push('configure', (core) => core.tokens.push(new Token('configure', 'configure', 0)))
      markdownIt.renderer.rules.configure = () => '<span>configure</span>'
    })
  }

  buildReplacers(): ComponentReplacer[] {
    return [new TestReplacer()]
  }

  buildTagNameAllowList(): string[] {
    return ['node-processor']
  }

  configureMarkdownItPost(markdownIt: MarkdownIt) {
    markdownIt.use(() => {
      markdownIt.core.ruler.push('post', (core) => core.tokens.push(new Token('post', 'post', 0)))
      markdownIt.renderer.rules.post = () => '<span>post</span>'
    })
  }
}
