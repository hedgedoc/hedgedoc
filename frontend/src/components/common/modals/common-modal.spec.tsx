/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { mockI18n } from '../../../test-utils/mock-i18n'
import { CommonModal } from './common-modal'
import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { Heart as IconHeart } from 'react-bootstrap-icons'

describe('CommonModal', () => {
  afterAll(() => {
    jest.resetAllMocks()
    jest.resetModules()
  })

  beforeAll(async () => {
    await mockI18n()
  })

  it('does not render if show is false', () => {
    const view = render(<CommonModal show={false}>testText</CommonModal>)
    expect(view.container).toMatchSnapshot()
  })

  it('renders correctly and calls onHide, when close button is clicked', async () => {
    const onHide = jest.fn()
    render(
      <CommonModal show={true} onHide={onHide} showCloseButton={true}>
        testText
      </CommonModal>
    )
    const modal = await screen.findByTestId('commonModal')
    expect(modal).toMatchSnapshot()
    const closeButton = await screen.findByRole('button')
    fireEvent(
      closeButton,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true
      })
    )
    expect(onHide).toHaveBeenCalled()
  })

  it('render correctly with title', async () => {
    render(
      <CommonModal show={true} titleI18nKey={'testTitle'}>
        testText
      </CommonModal>
    )
    const modal = await screen.findByTestId('commonModal')
    expect(modal).toMatchSnapshot()
  })

  it('render correctly with i18nTitle', async () => {
    render(
      <CommonModal show={true} titleI18nKey={'testTitle'}>
        testText
      </CommonModal>
    )
    const modal = await screen.findByTestId('commonModal')
    expect(modal).toMatchSnapshot()
  })

  it('render correctly with title icon', async () => {
    render(
      <CommonModal show={true} titleIcon={IconHeart}>
        testText
      </CommonModal>
    )
    const modal = await screen.findByTestId('commonModal')
    expect(modal).toMatchSnapshot()
  })

  it('render correctly with additionalClasses', async () => {
    render(
      <CommonModal show={true} additionalClasses={'testClass'}>
        testText
      </CommonModal>
    )
    const modal = await screen.findByTestId('commonModal')
    expect(modal).toMatchSnapshot()
  })

  describe('render correctly in size', () => {
    it('lg', async () => {
      render(
        <CommonModal show={true} modalSize={'lg'}>
          testText
        </CommonModal>
      )
      const modal = await screen.findByTestId('commonModal')
      expect(modal).toMatchSnapshot()
    })

    it('sm', async () => {
      render(
        <CommonModal show={true} modalSize={'sm'}>
          testText
        </CommonModal>
      )
      const modal = await screen.findByTestId('commonModal')
      expect(modal).toMatchSnapshot()
    })

    it('xl', async () => {
      render(
        <CommonModal show={true} modalSize={'xl'}>
          testText
        </CommonModal>
      )
      const modal = await screen.findByTestId('commonModal')
      expect(modal).toMatchSnapshot()
    })
  })
})
