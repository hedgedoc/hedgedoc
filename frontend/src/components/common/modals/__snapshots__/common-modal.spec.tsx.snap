// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonModal does not render if show is false 1`] = `<div />`;

exports[`CommonModal render correctly in size lg 1`] = `
<div
  class="modal-dialog modal-lg"
  data-testid="commonModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        <span />
      </div>
    </div>
    testText
  </div>
</div>
`;

exports[`CommonModal render correctly in size sm 1`] = `
<div
  class="modal-dialog modal-sm"
  data-testid="commonModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        <span />
      </div>
    </div>
    testText
  </div>
</div>
`;

exports[`CommonModal render correctly in size xl 1`] = `
<div
  class="modal-dialog modal-xl"
  data-testid="commonModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        <span />
      </div>
    </div>
    testText
  </div>
</div>
`;

exports[`CommonModal render correctly with additionalClasses 1`] = `
<div
  class="modal-dialog testClass"
  data-testid="commonModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        <span />
      </div>
    </div>
    testText
  </div>
</div>
`;

exports[`CommonModal render correctly with i18nTitle 1`] = `
<div
  class="modal-dialog"
  data-testid="commonModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        testTitle
      </div>
    </div>
    testText
  </div>
</div>
`;

exports[`CommonModal render correctly with title 1`] = `
<div
  class="modal-dialog"
  data-testid="commonModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        testTitle
      </div>
    </div>
    testText
  </div>
</div>
`;

exports[`CommonModal render correctly with title icon 1`] = `
<div
  class="modal-dialog"
  data-testid="commonModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        BootstrapIconMock_Heart
        Â 
        <span />
      </div>
    </div>
    testText
  </div>
</div>
`;

exports[`CommonModal renders correctly and calls onHide, when close button is clicked 1`] = `
<div
  class="modal-dialog"
  data-testid="commonModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        <span />
      </div>
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
    </div>
    testText
  </div>
</div>
`;
